From bca08980f0cdeed5efe60f415e6524a89c84a24a Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Tue, 7 Aug 2018 17:41:39 +0800
Subject: [PATCH 4311/5242] MLK-19152-2 drm/imx: lcdif: force 'mode_changed'
 when fb width changed

commit  ca7a163dfd6287ec05b531b2bf25c81f34f29d39 from
https://source.codeaurora.org/external/imx/linux-imx.git

In DRM atomic modeset check, it will not check the fb's width
change, so in later atomic commit, it will not disable the CRTC
which has no mode changed. But for LCDIF, the fb width related
registers configuration can not be done when LCDIF is running.
So force 'mode_changed' to be true when fb width changed.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
(cherry picked from commit 518ff82756a39ff2d2f750596295baa4f5fca4c5)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/lcdif/lcdif-plane.c |   11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/drivers/gpu/drm/imx/lcdif/lcdif-plane.c b/drivers/gpu/drm/imx/lcdif/lcdif-plane.c
index 15b015c..2045289 100644
--- a/drivers/gpu/drm/imx/lcdif/lcdif-plane.c
+++ b/drivers/gpu/drm/imx/lcdif/lcdif-plane.c
@@ -48,7 +48,9 @@ static int lcdif_plane_atomic_check(struct drm_plane *plane,
 	uint32_t bus_fmt;
 	struct lcdif_plane *lcdif_plane = to_lcdif_plane(plane);
 	struct lcdif_soc *lcdif = lcdif_plane->lcdif;
+	struct drm_plane_state *old_state = plane->state;
 	struct drm_framebuffer *fb = plane_state->fb;
+	struct drm_framebuffer *old_fb = old_state->fb;
 	struct drm_crtc_state *crtc_state;
 	struct drm_display_mode *mode;
 	struct drm_rect clip = { 0 };
@@ -92,6 +94,15 @@ static int lcdif_plane_atomic_check(struct drm_plane *plane,
 	if (!plane_state->visible)
 		return -EINVAL;
 
+	/* force 'mode_changed' when fb width changed, since
+	 * the pitch related registers configuration of LCDIF
+	 * can not be done when LCDIF is running.
+	 */
+	if (old_fb && likely(!crtc_state->mode_changed)) {
+		if (old_fb->width != fb->width)
+			crtc_state->mode_changed = true;
+	}
+
 	return 0;
 }
 
-- 
1.7.9.5

